source/blog/_posts/2018-05-30-pair-programming.md

```diff
@@ -0,0 +1,135 @@
+- Limitations (real or imagined)
+
+This picture may not be accurate but it's the context in which the dev actually makes changes to the code.
+
+Having two people keeping track of and building their context, instead of one, leads to many otherwise hard to find bugs being found in matter of minutes sometimes even instantly. That might sound overblown but I lost track of 
+how many time as soon a local test failed one of us went "Ahh that's the problem" and fixed something that was hard to see, like this:
+```php
+class Foo {
+
+    /**
+     * @var string
+    **/
+    private $prefix;
+
+    /**
+     * @var array
+    **/
+    private $config;
+
+    /**
+     * @var SomeService
+    **/
+    private $someService;
+
+    /**
+     * @var array
+    **/
+    private $stringVariantions;
+
+    public function __construct(array $config, SomeService $someService, string $prefix, string $str) 
+    {
+        $this->config = $config;
+        $this->stringVariantions = $this->createStringVariations($str);
+        $this->someService = $someService;
+        $this->prefix = $prefix;
+    }
+
+    private function createStringVariations(string $str): array
+    {
+       $arr = [];
+       for($i = 1; i< 5; $i++) {
+           
+           $arr[] = $prefix . $str . $i;
+       }
+       
+       return $out;
+    }
+}
+```
+
+A problem like this caused really weird behavior locally and in the failing tests, but my coworker spotted it right away:
+
+The lines `$this->stringVariantions = $this->createStringVariations($str);` and
+`$this->prefix = $prefix;` simply had to be swapped.
+
+This bug was an artifact left over from a round of refactoring that simply was not present in my mind because I worked on the `createStringVariations` method before running the tests.
+
+These things make pair programming itself enjoyable and productive for the developers actively participating. But lets switch gears and talk about the effects on the team. 
+
+## Islands made of knowledge
+As is often the case every member of our team knows some parts of the project better than others and this is fine. But when this specialisation develops too far, it leads to pretty big problems.
+
+In the extreme case the [bus factor](https://en.wikipedia.org/wiki/Bus_factor) drops to 1 and should somebody be absent no progress can be made.
+Even before the problem escalates that much, it will reduce the ability of the team to spread the work load properly.
+Because not every team member can handle every task, some team members will inevitably get more work assigned than others. This can stall progress while leaving a part of the team bored.
+
+Pair programming prevents this by simply involving two developers. The effects of that were obvious. My coworker and I both know the component we wrote well, and already briefed other team members on it.
+
+## Logistics & Agility
+Now for something that's well known: Most agile developers are not writing code all through their workday.
```

Comment:
"are not writing code nonstop/solely/exclusively"?
ich finde das so ziemlich unklar als erstes dachte ich du meintest "not writing code at all" bis ich gecheckt das du meinst auschließlich